GoogleSignIn/Sources/GIDAppCheck/Implementations/GIDAppCheck.m

```diff
@@ -52,45 +56,64 @@ - (BOOL)isPrepared {
   return [self.userDefaults boolForKey:kGIDAppCheckPreparedKey];
 }
 
-- (void)prepareForAppCheckWithCompletion:(nullable void (^)(FIRAppCheckToken * _Nullable,
-                                                            NSError * _Nullable))completion {
+- (void)prepareForAppCheckWithCompletion:(nullable GIDAppCheckCompletion)completion {
+  if (completion) {
+    @synchronized (self.prepareCompletions) {
+      [self.prepareCompletions addObject:completion];
+    }
+  }
+
+  @synchronized (self) {
+    if (self.preparing) {
+      return;
+    }
+
+    self.preparing = YES;
+  }
+
   dispatch_async(self.workerQueue, ^{
     if ([self isPrepared]) {
       NSError *error = [NSError errorWithDomain:kGIDAppCheckErrorDomain
                                            code:kGIDAppCheckAlreadyPrepared
                                        userInfo:nil];
-      if (completion) {
-        completion(nil, error);
+      @synchronized (self) {
+        for (GIDAppCheckCompletion savedCompletion in self.prepareCompletions) {
+          savedCompletion(nil, error);
+        }
+
+        [self.prepareCompletions removeAllObjects];
+        self.preparing = NO;
+
+        return;
       }
-      return;
     }
+
     [self.tokenFetcher limitedUseTokenWithCompletion:^(FIRAppCheckToken * _Nullable token,
                                                        NSError * _Nullable error) {
-      if (token) {
-        [self.userDefaults setBool:YES forKey:kGIDAppCheckPreparedKey];
-        if (completion) {
-          completion(token, nil);
+      @synchronized (self) {
+        NSError *maybeError = error;
+        if (!token && !maybeError) {
+          maybeError = [NSError errorWithDomain:kGIDAppCheckErrorDomain
+                                           code:kGIDAppCheckUnexpectedError
+                                       userInfo:nil];
         }
-        return;
-      }
-      if (error) {
-        if (completion) {
-          completion(nil, error);
+
+        if (!maybeError) {
+          [self.userDefaults setBool:YES forKey:kGIDAppCheckPreparedKey];
         }
-        return;
-      }
-      if (completion) {
-        NSError *noError = [NSError errorWithDomain:kGIDAppCheckErrorDomain
-                                               code:kGIDAppCheckUnexpectedError
-                                           userInfo:nil];
-        completion(nil, noError);
+
+        for (GIDAppCheckCompletion savedCompletion in self.prepareCompletions) {
+          savedCompletion(token, maybeError);
+        }
+
+        [self.prepareCompletions removeAllObjects];
+        self.preparing = NO;
```

Comment:
I think only access to `self.prepareCompletions` and `self.preparing` needs to be synchronzied.